File: 프로그래머스/3/42893. 매칭 점수/매칭 점수.js
function solution(word, pages) {
  const n = pages.length;
  const target = word.toLowerCase();

  const urlRe = /<meta[^>]*property="og:url"[^>]*content="([^"]+)"/;
  const linkRe = /<a href="([^"]+)"/g;

  const urls = Array(n).fill("");
  const outLinks = Array.from({ length: n }, () => []);
  const base = Array(n).fill(0);
  const linkScore = Array(n).fill(0);

  const urlToIndex = new Map();

  const countWordOccurrences = (htmlLower, targetLower) => {
    let cnt = 0;
    let cur = "";

    for (let i = 0; i < htmlLower.length; i++) {
      const c = htmlLower[i];
      if (c >= "a" && c <= "z") {
        cur += c;
      } else {
        if (cur.length) {
          if (cur === targetLower) cnt++;
          cur = "";
        }
      }
    }
    if (cur.length && cur === targetLower) cnt++;
    return cnt;
  };

  // 1) parse url / outlinks / base score
  for (let i = 0; i < n; i++) {
    const html = pages[i];

    // url
    const um = html.match(urlRe);
    if (um) {
      urls[i] = um[1];
      urlToIndex.set(urls[i], i);
    }

    // outlinks
    linkRe.lastIndex = 0; // reset global regex
    let m;
    while ((m = linkRe.exec(html)) !== null) {
      outLinks[i].push(m[1]);
    }

    // base score (word boundary by non-letters)
    const lower = html.toLowerCase();
    base[i] = countWordOccurrences(lower, target);
  }

  // 2) distribute link score
  for (let i = 0; i < n; i++) {
    const outCnt = outLinks[i].length;
    if (outCnt === 0) continue;

    const share = base[i] / outCnt;
    for (const to of outLinks[i]) {
      if (urlToIndex.has(to)) {
        linkScore[urlToIndex.get(to)] += share;
      }
    }
  }

  // 3) pick best matching score (tie -> smallest index)
  let bestIdx = 0;
  let bestScore = base[0] + linkScore[0];

  for (let i = 1; i < n; i++) {
    const s = base[i] + linkScore[i];
    if (s > bestScore) {
      bestScore = s;
      bestIdx = i;
    }
  }

  return bestIdx;
}

File: 프로그래머스/3/42893. 매칭 점수/매칭 점수.py
import re

def solution(word, pages):
    target = word.lower()
    n = len(pages)

    url_re = re.compile(r'<meta[^>]*property="og:url"[^>]*content="([^"]+)"')
    link_re = re.compile(r'<a href="([^"]+)"')

    urls = [""] * n
    out_links = [[] for _ in range(n)]
    base = [0.0] * n
    link_score = [0.0] * n

    url_to_idx = {}

    def count_word_occurrences(html: str, target_lower: str) -> int:
        s = html.lower()
        cnt = 0
        cur = []
        for ch in s:
            if 'a' <= ch <= 'z':
                cur.append(ch)
            else:
                if cur:
                    if ''.join(cur) == target_lower:
                        cnt += 1
                    cur.clear()
        if cur and ''.join(cur) == target_lower:
            cnt += 1
        return cnt

    # 1) parse url / outlinks / base score
    for i, html in enumerate(pages):
        m = url_re.search(html)
        if m:
            urls[i] = m.group(1)
            url_to_idx[urls[i]] = i

        out_links[i] = link_re.findall(html)
        base[i] = float(count_word_occurrences(html, target))

    # 2) distribute link score
    for i in range(n):
        out_cnt = len(out_links[i])
        if out_cnt == 0:
            continue
        share = base[i] / out_cnt
        for to in out_links[i]:
            j = url_to_idx.get(to)
            if j is not None:
                link_score[j] += share

    # 3) best matching score (tie -> smallest index)
    best_idx = 0
    best = base[0] + link_score[0]
    for i in range(1, n):
        s = base[i] + link_score[i]
        if s > best:
            best = s
            best_idx = i

    return best_idx